### PR TITLE
Reduces the number of tests to run in codec_impl_test.cc

### PR DESCRIPTION
**Pull request by soya3129**
Signed-off-by: Yang Song <yasong@google.com>
*Description*: This PR includes two changes: (1) randomly select settings values to reduce the number of tests to run. (2) change the local variable buf to a static variable buf_tmp. 
*Risk Level*: Low. Test only change.
*Testing*: n/a.
*Docs Changes*: n/a
*Release Notes*: Test only change.

### CODE REVIEW

File: test/common/http/http2/codec_impl_test.cc [hunk 1]
```diff
@@ -815,15 +828,24 @@ INSTANTIATE_TEST_CASE_P(Http2CodecImplTestDefaultSettings, Http2CodecImplTest,
                         ::testing::Combine(HTTP2SETTINGS_DEFAULT_COMBINE,
                                            HTTP2SETTINGS_DEFAULT_COMBINE));
 
+namespace {
+// Randomly picks a testing value to reduce the number of tests to run.
```
Comment (mattklein123) [inline test/common/http/http2/codec_impl_test.cc]: Generally not a fan of random tests. What's the reason for this change?

File: test/common/http/http2/codec_impl_test.cc [hunk 2]
```diff
@@ -758,7 +769,9 @@ class Http2CodecImplStreamLimitTest : public Http2CodecImplTest {};
 // TODO(PiotrSikora): add tests that exercise both scenarios: before and after receiving
 // the HTTP/2 SETTINGS frame.
 TEST_P(Http2CodecImplStreamLimitTest, MaxClientStreams) {
+  ENVOY_LOG_MISC(error, "Client http2 settings: ");
```
Comment (mattklein123) [inline test/common/http/http2/codec_impl_test.cc]: Is this needed at error level? Can you just run the test with debug logging to see (and print at that level if not already)?

File: test/common/http/http2/codec_impl_test.cc [hunk 3]
```diff
@@ -118,12 +130,11 @@ class Http2CodecImplTest : public testing::TestWithParam<Http2SettingsTestParam>
...
       return;
     }
-    uint8_t buf[METADATA_MAX_PAYLOAD_SIZE] = {0};
-    data.copyOut(0, length, static_cast<void*>(buf));
+    data.copyOut(0, length, static_cast<void*>(buf_tmp));
     data.drain(length);
     // Keeps the frame header (9 bytes) valid, and corrupts the payload.
-    buf[10] |= 0xff;
-    data.add(buf, length);
+    buf_tmp[10] |= 0xff;
```
Comment (mattklein123) [inline test/common/http/http2/codec_impl_test.cc]: Not sure I understand this change?

File: test/common/http/http2/codec_impl_test.cc [hunk 4]
```diff
@@ -102,12 +107,19 @@ class Http2CodecImplTest : public testing::TestWithParam<Http2SettingsTestParam>
         }));
   }
 
+  // Using error level logging so the settings values can be logged.
```
Comment (mattklein123) [inline test/common/http/http2/codec_impl_test.cc]: Same comment about logging level

File: test/common/http/http2/codec_impl_test.cc (hunk 1 above)
Comment (soya3129) [reply test/common/http/http2/codec_impl_test.cc]: The test hits memory limit when runs in Forge (an internal system to run tests). But it runs fine with bazel. Maybe that means I can make the change in google3 only (?). I hoped the continuous build can cover all the settings combinations.

Currently, each test runs 255 times with different settings. Using randomly selected values, each test runs once. But I am totally OK to revert this change, :)

Comment (mattklein123) [reply test/common/http/http2/codec_impl_test.cc]: @soya3129 can you chat with @htuch and @alyssawilk about the internal Google issue and circle back here?

File: test/common/http/http2/codec_impl_test.cc (hunk 2 above)
Comment (soya3129) [reply test/common/http/http2/codec_impl_test.cc]: Because the settings values are randomly selected, if a test fails, using error level logging means we know which settings combination triggers the error even if the test doesn't run with debug binary. For example, if the test fails in continuous opt build, we can easily reproduce the failure by setting the values in the log.

File: test/common/http/http2/codec_impl_test.cc (hunk 3 above)
Comment (soya3129) [reply test/common/http/http2/codec_impl_test.cc]: To make buf_tmp a static member, we only need to allocate the memory once. I think, previously, we may need to allocate the memory every time the function is called.

File: test/common/http/http2/codec_impl_test.cc (hunk 4 above)
Comment (soya3129) [reply test/common/http/http2/codec_impl_test.cc]: Please see the reply above.

File: test/common/http/http2/codec_impl_test.cc (hunk 1 above)
Comment (htuch) [reply test/common/http/http2/codec_impl_test.cc]: I think the main thing is we don't need to test with the Cartesian product, but can pick a baseline configuration and then do sensitivity analysis in the tests one parameter at a time. I would do that for the tests in this file, and if we one day we find a regression with more complex parameters, we can consider making this more complicated.

Comment (mattklein123) [reply test/common/http/http2/codec_impl_test.cc]: Agreed. I still don't understand why this fails internally due to memory issues, that doesn't make any sense to me. But I probably don't want to know.

Mainly, I don't like random. Can we do min/max/mid testing or something like that across all the settings? That should greatly reduce the number of tests and allow us to revert the rest of these changs?

Comment (soya3129) [reply test/common/http/http2/codec_impl_test.cc]: OK! Will create another PR to reduce cases by running min/max/mid testing. Thanks for the reviews and discussion!
